### DISCUSSION TITLE
Way of "cleaning" a file without saving it

### DISCUSSION THREAD

**Discussion by dantard** [pymupdf/PyMuPDF#3531] [Looking for help]
Hello again, 
I was looking for a way to "clean" a pdf file in memory, that is, without saving it. I found this workaround but I am not sure it is the best way because sometimes is slow and additionally I found a problem with it:

```
document = pymupdf.open("file.pdf")
clean = document.tobytes(encryption=PDF_ENCRYPT_KEEP, deflate=True, garbage=3)
document = pymupdf.open("pdf", clean)
```

The problem is that with [this](https://www.irs.gov/pub/irs-pdf/i1040tt.pdf) file, that I randomly found online, the `tobytes` gets stuck. Not sure it is the only file with which this would happen.

Thanks in advance.

**Comment by JorjMcKie** [comment]
Your approach is correct!
If you want to compress, always look at all the details. For example using object streams compression may in fact take longer if the PDF is large and did not use them before.
The more compression you want / request, the higher the chance that things take longer - not a surprise really, is it?
That example file indeed has problems - but they are not related to the `.tobytes()` method: also occurs when using save.
